### ISSUE TITLE
Course creator can list all users

### ISSUE THREAD

**Issue by michzimny** [up2university/docker-moodle#52]
As a regular user with the course creator role in a category, I can list all the users with their email addresses. I go to a course I've created and choose "Enrolled users"->"Enrol users". 

Can we disable at all listing users not assigned to a particular course or at least users not belonging to course creator's category?

**Comment by nadavkav** [comment]
There is not a option for this in Moodle, as far as I can tell.
Moodle enrollment dialog always show all the not enrolled into the current course user list in the entire system. when using manual enrollment.
If we turn it off and only use cohort enrollment... then we have to add students to cohorts and not to courses. which is a different flow (than the one in which a teacher adds users to the course on he/she own)
We can also disable the option for a teacher to manually enrol user into the course, and only let him/her create an easy enrollment key. will that be good?

Is this issue is raised as part of a clear GDPR restriction  ?

**Comment by michzimny** [comment]
I just asked our work package that is in charge of such issues to make a decision whether it is a problem for us or not.

To my knowledge, GDPR requires to design solutions in such a way that the user data is processed (e.g. presented to someone) only if it is really needed. I think showing email addresses to teachers from other countries does not comply with that. But maybe we are able to defend ourselves by some of many exceptions to that rule.

**Comment by nadavkav** [comment]
To make it a little bit more clear, just hiding the email in most places and maybe all around the system is something we can do, but not the ability to see user's names when looking for enrolling new students into courses. (unless, on a later stage of the project we move entirely to enrollment based on external SIS platforms, with no manual course enrollments)
Searching though the admin settings for "showuseridentity", you will see that email is displayed by default, but we can hide it in most places, regulated by core. (not considering 3rd party plugins)

**Comment by michzimny** [comment]
I think that hiding email addresses would be a good improvement. However let's wait for the opinion of our GDPR experts. I can be wrong, and the current state might also be fine.

**Comment by nadavkav** [comment]
Also, let's wait for the tools Moodle HQ will release in a few weeks. as it
should be solving some if not all of our issues and concerns.

On Tue, Mar 6, 2018 at 11:36 AM, Michał Zimniewicz <notifications@github.com

**Comment by michzimny** [comment]
We discussed this issue during MT call today. Peter asked to disable the option to manually enrol users into the course, and only allow create enrollment keys.

Are there any other places where our users (excluding us, the administrators) could see the list of all users in Moodle?

**Comment by nadavkav** [comment]
BTW, going through security issues DB:
https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-15110

Maybe we should link our github repo to https://snyk.io/

**Comment by michzimny** [comment]
> Peter asked to disable the option to manually enrol users into the course, and only allow create enrollment keys.

How about this? Can we do this?

cc @athird

**Comment by nadavkav** [comment]
We can disable manual enrollment on a system level (as admin), so none of the courses can use it.
but we can not do it for only a few course. it's all or nothing. hope it is what we need.
https://learn.test.up2university.eu/admin/settings.php?section=manageenrols
(Click to "eye" to disable manual enrollment )

**Comment by michzimny** [comment]
I think we want to disable manual enrollment in all courses. 

@athird, could we appropriately set this option to all instances?

**Comment by michzimny** [comment]
It works fine in dev.

@athird, can we deploy to other instances?

**Comment by michzimny** [comment]
@athird?

**Comment by athird** [comment]
@michzimny Yes, I'll merge it, sorry!

**Comment by michzimny** [comment]
Ok, thanks. I'm going to deploy it to both pedagogy and production in the evening.

**Comment by michzimny** [comment]
Reopening because of #81.

**Comment by michzimny** [comment]
Could we look for a solution for this please?

**Comment by nadavkav** [comment]
You could remove any capability that allow a "Course creator" role to view users, by updating the legacy role "coursecreator" 
![image](https://user-images.githubusercontent.com/231168/46909048-bfec8d00-cf34-11e8-9bbf-478ba70a1bd3.png)
Use the filter input box to narrow down anything that is user related by inputting "user"
Click off any "allow" checkbox that seems relevant.
Or..
Create a new role "Course creator - Privacy" and remove those capabilities, and use that role, instead of the legacy one.

**Comment by michzimny** [comment]
@athird and @Tabbasum - could you please have a look and try to apply this suggestion?

**Comment by athird** [comment]
@nadavkav The course creator role doesn't currently have any capabilities that mention "user", so there's nothing to try. Do you have any further suggestions?

Also, what does the course creator user have to do in order to see the list of all users?

**Comment by michzimny** [comment]
> Also, what does the course creator user have to do in order to see the list of all users?

Assuming that question is for me. As course creator I was able to create course, then go to the course and choose "Enrolled users"->"Enrol users" to see all users in the system with their names and email addresses.

@athird, is it what are you looking for?

**Comment by athird** [comment]
@michzimny Yes, thanks, that's what I was looking for. I thought that option wouldn't be visible if manual user enrolment was disabled, though. 

So could the right fix be to remove the "enrol" capabilities (except self-enrol) from Course Creators?

**Comment by michzimny** [comment]
My tests have shown that disabling manual user enrolment makes course creators unable to enrol to their newly created courses. So, one was able to create a course, but then he or she had no permissions in that course.

Nadav have then proposed other suggestions that need to be configured and tested.

**Comment by athird** [comment]
But that was disabling the whole manual enrolment functionality, wasn't it? Looking at the screenshot [here](https://github.com/up2university/docker-moodle/issues/81#issuecomment-395099248), though, if we edit capabilities directly, it draws a distinction between self-enrolment and enrolling other users. So we could leave the first enabled and disabled the others.

**Comment by michzimny** [comment]
Ah, ok. So yes, it might work as you're saying...

**Comment by athird** [comment]
Just looked, I don't think it will work. The capability for self-enrolment is related to allowing any users to pick a course and enrol themselves in it.

But now I'm more deeply confused. Looking on learn.test..., the course creator role doesn't *have* any of the enrol capabilities ticked, so how are they able to do it? Or does the teacher role have them, and actually creators are getting the capabilities because they're automatically teachers?

**Comment by athird** [comment]
@michzimny I've removed some of the enrolment capabilities from teachers. Would you be able to check if it's made any difference to the "list all users" issue? My various accounts are all either too low or too high privilege to test this easily. :-)

**Comment by msbachler** [comment]
I found this: https://moodle.org/mod/forum/discuss.php?d=357154
It seem to describe how to change the roles and permissions so emails can't be seen. Just investigating how we could set the two changes he suggests programmatically.

**Comment by msbachler** [comment]
Actually reading your issue again, I am not sure you want to turn off seeing emails displayed beside user in lists completely, as the above would do. Can someone clarify if this is what you want to do or not?

**Comment by msbachler** [comment]
Allan has clarified the requirements for this issue, so ignore the above 2 entries.

**Comment by michzimny** [comment]
I think what we want to do is to disable listing all users in the system. A suggestion was that it could be achieved by disabling manual enrolments to courses (as we enrol users via enrolment codes). But the solution cannot prevent non-admin users from creating courses.

**Comment by Callustian** [comment]
@michzimny @msbachler  It is possible to change the settings of a category. Deleting all elements (student, teacher, master, ...) of "HomeCategory: <category> categoryPermissions View Participants" as Administrator forbids a user to list the participants. I think this answers the requirements.
